### PR TITLE
Add encode method to convert Sections into GPP String

### PR DESCRIPTION
**Pull request by duduchristian**

### CODE REVIEW

File: util/encoding.go [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+)
+
+var (
+	fibEncodeNumOutOfRangeErr = fmt.Errorf("the number to be encoded is out of range")
+)
+
+// enlarge enlarges the underlying byte slice.
+func (bs *BitStream) enlarge(n uint16) {
+	final := (bs.p + n + 7) / 8
+	for final > bs.Len() {
```
Comment (hhhjort) [inline util/encoding.go]: It would probably by more efficient to append one mass of bytes, rather than looping over an append statement like this. 
`
size := final - bs.Len()
chuck := make([]byte, size, size)
bs.b = append(bs.b, chunk...)
`
The new array created by make should contain all zeros in all the elements of the array. The documentation isn't 100% clear, but that is how it reads to me.

This may not be a big difference when enlarge is used by the write functions below, as it will only be writing one or two bytes at a time. So I am not too worried about the final form of enlarge currently, but may be more of a factor if we make more expansive write functions in the future. In particular `WriteIntRange()` could probably calculate the total size needed fairly quickly, and then ensure the byte slice in the bit stream has sufficient capacity in one step, rather than depending on a sequence of `enlarge()` calls.

File: encode.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+	gppType    byte = 0x3
+	gppVersion byte = 0x1
+)
+
+var (
+	duplicatedSectionErr = fmt.Errorf("duplicated sections")
+)
+
+func Encode(sections []Section) (string, error) {
+	bs := util.NewBitStream(nil)
```
Comment (hhhjort) [inline encode.go]: It may be more efficient to estimate the final size of the bitstream, and then create a `[]byte` with that capacity. That will save a lot of reallocations and copies as the bitstream is expanded with append statements. Most applications are not so memory constrained that allocating a handful of bytes more than needed per GPP string should be an issue.

**Comment by hhhjort** [comment]
Can you put a note in the README that, in practice, only CMPs should be writing GPP strings as per the IAB documentation? We don't want to encourage people to be generating GPP strings in production when they really should not be.

**Comment by hhhjort** [comment]
Just a friendly ping to see if you have read the comments, and intend to keep working on this PR.

Thanks

**Comment by duduchristian** [comment]
Coming back, still working on this.

> Just a friendly ping to see if you have read the comments, and intend to keep working on this PR.
> 
> Thanks

Coming back, still working~

File: encode.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+			return "", duplicatedSectionErr
+		}
+		if prevID+1 == id {
+			intRange.Range[len(intRange.Range)-1].EndID = uint16(id)
+		} else {
+			intRange.Range = append(intRange.Range, util.IRange{StartID: uint16(id), EndID: uint16(id)})
+		}
+		prevID = id
+	}
+	intRange.Size = uint16(len(sections))
```
Comment (hhhjort) [inline encode.go]: The size should be `len(intRange.Range)`, not the total number of sections.

File: encode_test.go [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+	"testing"
+)
+
+type gppEncodeTestData struct {
+	description string
+	sections    []Section
+	expected    string
+}
+
+var testData = []gppEncodeTestData{
```
Comment (hhhjort) [inline encode_test.go:23]: Should probably have a test or two where multiple sections are encoded.

File: util/encoding_test.go [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+		bs := NewBitStream(nil)
+		p := 0
+		for p < len(c.data)*8 {
+			p = writeNBits(bs, c.data, p, 1)
+		}
+		assert.Equal(t, c.data, bs.b, fmt.Sprintf("test case %d [%s] failed", no, c.description))
+	}
+}
+
+func TestWriteByte2(t *testing.T) {
```
Comment (hhhjort) [inline util/encoding_test.go:111]: The way you have your tests set up, you will always be writing on the even boundaries of your bit units. For example, `TestWriteByte2` will always be writing to even bit indexes. `TestWriteByte8` will always be writing on byte boundaries. You need to mix in some arbitrary offsets to exercise the tests to ensure the functions work properly at arbitrary indexes.

Comment (duduchristian) [reply util/encoding_test.go:111]: refer to TestArbitraryWrite

**Comment by hhhjort** [comment]
It may be interesting to write a test where we decode some GPP strings, then re-encode them and compare to see if we got the original string back. It could uncover a subtle error, or potentially that the spec is under specified such that there are multiple ways to encode a set of data into a string.

File: encode.go [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+)
+
+const (
+	// the first 6 bits of the header must always evaluate to the integer '3'.
+	gppType    byte = 0x3
+	gppVersion byte = 0x1
+)
+
+var (
+	duplicatedSectionErr = fmt.Errorf("duplicated sections")
```
Comment (bsardo) [inline encode.go]: Nitpick: when there aren't any parameters in your error message, I believe it is preferable to use `errors.New` from the `errors` package.

File: util/encoding.go [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+import (
+	"encoding/base64"
+	"fmt"
+	"unsafe"
+)
+
+var (
+	fibEncodeNumOutOfRangeErr = fmt.Errorf("the number to be encoded is out of range")
+	fibEncodeInvalidRange     = fmt.Errorf("the range is invalid")
+)
```
Comment (bsardo) [inline util/encoding.go]: Nitpick: when there aren't any parameters in your error message, I believe it is preferable to use `errors.New` from the `errors` package.

File: encode.go [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+	err := bs.WriteIntRange(intRange)
+	if err != nil {
+		return "", fmt.Errorf("write int range error: %v", err)
+	}
+
+	builder.Write(bs.Base64Encode())
+
+	for _, sec := range sections {
+		// TODO: add a parameter to decide whether a GPC segment should be included.
```
Comment (bsardo) [inline encode.go]: I suggest we use another method to track this instead of putting TODOs in the code. If this is something that needs to be done in a separate PR, perhaps we can track this in a different way. Maybe create an issue? @hhhjort thoughts?

I noticed the `Encode` methods for some of the regs has code to exit early if this value is false but it doesn't look like there is currently any way to take that code path so tests should probably be added for those lines when this TODO is addressed.

File: encode_test.go [hunk 9]
```diff
@@ -0,0 +1,356 @@
...
+				MspaCoveredTransaction:  2,
+				MspaOptOutOptionMode:    1,
+				MspaServiceProviderMode: 2,
+			},
+			GPCSegment: sections.CommonUSGPCSegment{
+				SubsectionType: 1,
+				Gpc:            false,
+			},
+			SectionID: constants.SectionUSPCA,
+			Value:     "BlgWEYCY.QA"},
```
Comment (bsardo) [inline encode_test.go]: Nitpick: can you put `uspca.USPCA{` on the next line and indent so all of the sections line up so it is easier to read? The same applies to other test cases you have here.

File: util/encoding_test.go [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+			rest := N - p
+			for choices[len(choices)-1] > rest {
+				choices = choices[:len(choices)-1]
+			}
+			p = writeNBits(bs, c.data, p, choices[rand.Intn(len(choices))])
+		}
+
+		assert.Equal(t, c.data, bs.b, fmt.Sprintf("test case %d [%s] failed", no, c.description))
+	}
+}
```
Comment (bsardo) [inline util/encoding_test.go]: I understand that there is value here as this test can be used to verify writes off of byte boundaries, however, I think there is a better way to handle this. The exact tests executed are unpredictable given the use of the random generator. I suggest deleting this test and instead specifying the exact off byte boundary test cases you want to cover.
If you follow my approach suggested in another comment that asks for the test cases to be declared in each test rather than declared globally in the file, you can easily add additional test cases to each of the write function tests to handle these off byte boundary scenarios.
It's a bit more verbose but predictable and easier to maintain since each test is isolated.

**Comment by hhhjort** [comment]
Sorry we have been a bit busy and haven't responded in a bit. This PR seems to be in a good state, and we should get the final review in soon.

File: test.sh
Comment (bsardo) [inline test.sh:1]: It appears you want this file to be executable however it it is read-only.

Also, we should probably update the `.gitignore` to include `release/` since you're creating files in that directory when this script is executed.

Comment (bsardo) [inline test.sh:1]: Can you also add `.idea/` and `.vscode/` to `.gitignore`?

File: util/encoding_test.go [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+		{[]uint16{1}, []byte{0xc0}, nil},
+		{[]uint16{2}, []byte{0x60}, nil},
+		{[]uint16{4}, []byte{0xb0}, nil},
+		{[]uint16{5}, []byte{0x18}, nil},
+		{[]uint16{7, 12}, []byte{0x5d, 0x60}, nil},
+		{[]uint16{0, 6765}, nil, fibEncodeNumOutOfRangeErr},
+		{[]uint16{6764}, []byte{0x55, 0x55, 0x60}, nil},
+	}
+
+	for _, test := range fibWriteTestData {
```
Comment (bsardo) [inline util/encoding_test.go:407]: Can you please add a `name` field to each test case describing what the test case covers and make use of `t.Run`:
```
for name, test := range tests {
		t.Run(name, func(t *testing.T) {
```

File: util/encoding_test.go [hunk 12]
```diff
@@ -0,0 +1,235 @@
...
+		},
+		{
+			[]IRange{{2, 5}, {6, 8}, {13, 16}, {28, 9999}},
+			nil,
+			fibEncodeNumOutOfRangeErr,
+		},
+	}
+
+	var err error
+	for _, c := range cases {
```
Comment (bsardo) [inline util/encoding_test.go]: Same comment as above. Can you please add a `name` field to each test case describing what the test case covers and make use of `t.Run`:
```
for name, test := range tests {
		t.Run(name, func(t *testing.T) {
```

File: util/encoding_test.go [hunk 13]
```diff
@@ -0,0 +1,235 @@
...
+		bs := NewBitStream(nil)
+		intRange := IntRange{Range: c.r, Size: uint16(len(c.r))}
+		err = bs.WriteIntRange(&intRange)
+		assert.Equal(t, c.err, err)
+		if err == nil {
```
Comment (bsardo) [inline util/encoding_test.go]: Nitpick: can you make this condition based on the expectation `c.err` instead of `err`? Actually, why is this condition needed at all? Shouldn't we know what the value should be given the test inputs regardless of whether there is an error?

File: util/encoding.go [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+}
+
+// WriteTwoBitField .
+func (bs *BitStream) WriteTwoBitField(bList []byte) {
+	for _, b := range bList {
+		bs.WriteByte2(b)
+	}
+}
+
+// WriteFibonacciInt .
```
Comment (bsardo) [inline util/encoding.go]: Some of the functions here only have the name of the function as a comment which I believe was done temporarily to quiet the lint checker. Can you please expand on these comments? IMO functions like this one could especially use more information to inform the reader of intent.

File: util/encoding_test.go [hunk 15]
```diff
@@ -0,0 +1,235 @@
...
+		"Test 2",
+		[]byte("I shall dedicate myself to the interests of the country in life and death"),
+	},
+	{
+		"Test 3",
+		testData,
+	},
+}
+
+func TestWriteByte1(t *testing.T) {
```
Comment (bsardo) [inline util/encoding_test.go:78]: Can we maybe rewrite this and the other write byte tests so they are isolated making use of `t.Run` with their own test cases (which can explicitly call out byte boundary scenarios) instead of shared ones and explicitly call the function that is under test instead of making a helper function with conditional logic determine what to call? IMO this would make the tests more straightforward and simplify the helper.

How about something like this:
```
func TestWriteByte1(t *testing.T) {
	testCases := []struct{
		name string
		data []byte
	}{
		{
			name: "single_byte",
			data: []byte("a"),
		},
		{
			name: "bytes_string",
			data: []byte("too young too simple sometimes naive"),
		},
		{
			name: "bytes_unicode",
			data: []byte{0x04, 0xa2, 0x03, 0xb1, 0x00, 0x2b},
		},
	}

	for _, tc := range testCases {
		t.Run(tc.name, func(t *testing.T) {
			bs := NewBitStream(nil)
			p := 0
			for p < len(tc.data)*8 {
				dataToWrite := prepareNBits(tc.data, p, 1)
				bs.WriteByte1(byte(dataToWrite))
				p += 1
			}
			assert.Equal(t, tc.data, bs.b)
		})
	}
}
```

In this case I changed `writeNBits` to `prepareNBits`.

**Comment by bsardo** [comment]
Hi @duduchristian, just checking in to see if you have any time to address my comments. Overall, it looks good. No rush and thanks!

**Comment by duduchristian** [comment]
> 

yep, I get some time next week. Just coming back from my holiday hahah.

**Comment by bsardo** [comment]
Hi @duduchristian, just checking in to see if you have some time to address the outstanding comments. No rush 🙂

**Comment by bsardo** [comment]
Hi @duduchristian. It looks like you made some of the updates but there are a few more comments outstanding. Are you planning to address those as well? Thanks!

**Comment by duduchristian** [comment]
> Hi @duduchristian. It looks like you made some of the updates but there are a few more comments outstanding. Are you planning to address those as well? Thanks!

Sorry I was so busy in the previous weeks. It seems I've addressed all the problems mentioned in comments, can you check the code and if there are any problems, make comments just as what you did before. Thank you.

File: .deepsource.toml [hunk 16]
```diff
@@ -0,0 +1,7 @@
+version = 1
+
+[[analyzers]]
+name = "go"
+
+  [analyzers.meta]
+  import_root = "github.com/duduchristian/go-gpp"
```
Comment (hhhjort) [inline .deepsource.toml]: Is this file supposed to be included? If so, should the `import_root` not be modified to point to this repository?

File: util/encoding.go [hunk 17]
```diff
@@ -0,0 +1,243 @@
...
+	}
+
+	const expThreshold = 64
+	const step = 16
+	var temp []byte
+
+	if final <= expThreshold {
+		finalF := float32(final - 1)
+		// Find out the leftmost one's position.
+		pos := *(*int)(unsafe.Pointer(&finalF))>>23&255 - 127
```
Comment (hhhjort) [inline util/encoding.go]: I am not sure what this is calculating or what expThreshold is supposed to be?

Comment (hhhjort) [reply util/encoding.go]: Would it be better to say `size := cap(bs.b) * 2` rather than using this complex function that is creating an int pointer out of a float pointer? This would seem to depend strongly on just how the CPU architecture represents ints and floats in memory. 

Actually you would need to loop in case you need more than one doubling in size.
```suggestion
		if final <= expThreshold {
		        size := expThreshold
		        for size > final*2 {
		                size = size/2
		        }
		        temp = make([]byte, final, size)
```
This would also fix the bitstream starting at some odd size.

File: util/encoding.go [hunk 18]
```diff
@@ -0,0 +1,243 @@
...
+		size := 1
+		if pos >= 0 {
+			size = 1 << (pos + 1)
+		}
+		temp = make([]byte, final, size)
+	} else {
+		size := (final-expThreshold+step-1)/step*step + expThreshold
```
Comment (hhhjort) [inline util/encoding.go]: Might want to add a parenthesis to make the order of operations clearer for humans.
`((final-expThreshold+step-1)/step)*step`
It took me a bit to parse this correctly in my head.

File: util/encoding.go [hunk 19]
```diff
@@ -0,0 +1,243 @@
...
+
+func NewBitStreamForWrite() *BitStream {
+	return NewBitStream(getByteSlice())
+}
+
+// enlarge the underlying byte slice.
+// In short, if the final size is less than the expThreshold, double the size of the underlying byte array,
+// otherwise, enlarge the size of the array by a specified step.
+func (bs *BitStream) enlarge(n uint16) {
+	final := int(bs.p+n+7) / 8
```
Comment (hhhjort) [inline util/encoding.go:26]: This is calculating the final position (or size) from where the bitstream pointer currently is. It would be sketchy to write to the middle of a bit stream, but should probably make this clearer in the comment so someone doesn't assume it can divine where the end of the bit stream is.

File: util/encoding.go [hunk 20]
```diff
@@ -0,0 +1,243 @@
...
+According to the algorithm, no sequential ones can be found in the bit sequence, and the final bit is always a single 1,
+which makes it easy to extract several continuous Fibonacci encoded integers out from a bit stream if we
+place an additional 1 in the end, since 11 could be an end-of-sequence indicator. The advantage of this is,
+that a sequence of numbers can be encoded into a sequence of bits without the need to know the length of the
+bit sequences in advance.
+*/
+func (bs *BitStream) WriteFibonacciInt(num uint16) error {
+	// The num should be [1,6765). Actually once the num is larger than or equal to 987,
+	// the efficiency of Fibonacci Encoding would be no better than 'WriteUint16'.
+	if num <= 0 || num >= fibonacci(fibLen) {
```
Comment (hhhjort) [inline util/encoding.go]: If you are limiting to `fibLen`, then you can just lookup in the `fibLookup[i]` array rather than calling the `fibonacci()` function. The function just adds code to handle the case where the number if greater than `fibLen`.

File: util/encoding.go [hunk 21]
```diff
@@ -0,0 +1,243 @@
...
+
+The basic idea is 12 bits for size, 1 bit indicating if a range refers to a single number or not, followed by
+one or two Fibonacci encoded numbers, and then followed by other ranges fell behind.
+
+For example, an IntRange consisting of [1, 1], [3, 4] would be encoded into 000000000010 0 11 1 0011 11.
+There are two ranges, so the size is encoded into 000000000010. The first range [1, 1] refers to a single integer 1,
+so the indicator is 0, followed by Fibonacci encoded 1, while the next range [3, 4] refers to a range containing
+not only one integer, so the indicator is 1, followed by Fibonacci encoded 3 and 1 (4-3).
+*/
+func (bs *BitStream) WriteIntRange(intRange *IntRange) error {
```
Comment (hhhjort) [inline util/encoding.go:214]: Shouldn't Fibonacci enter the name someplace as the GPP spec also defines a non-Fibonacci int range. The non-Fibonacci range will be needed once TCFv2 support is added.

**Comment by bsardo** [comment]
Hi @duduchristian, thanks for responding to our comments. We recently left a few more for you to address when you get a chance. If you don't mind, could you please merge with master instead of force pushing? It will save us a lot of time reviewing. Thanks!

File: util/encoding.go [hunk 22]
```diff
@@ -31,23 +30,18 @@ func (bs *BitStream) enlarge(n uint16) {
...
+	doubleCap := newCap + newCap
+	if final > doubleCap {
+		newCap = final
 	} else {
-		size := (final-expThreshold+step-1)/step*step + expThreshold
-		temp = make([]byte, final, size)
+		if newCap < 32 {
+			newCap = doubleCap
+		} else {
+			newCap += newCap / 4
```
Comment (hhhjort) [inline util/encoding.go]: This requires some work.

For example, if final = 70 and newCap = 40, then `final > doubleCap` is false, `newCap < 32` is false, so we have `newCap += newCap / 4` which makes  the new newCap = 50. The `make` will then fail as you try to make an array sized larger than its capacity.

File: util/encoding.go (hunk 21 above)
Comment (hhhjort) [reply util/encoding.go:214]: Would you like to comment on the difference between the Fibonacci Int range and the legacy Int Range that is used in TCF. We do plan to add TCFv2 code to this library once we get some time to do it. It hasn't been too urgent as an independent TCFv2 library exists.

**Review by hhhjort — DISMISSED** [review]
Looks good to me now. Will get a second pair of eyes to review before we merge.

File: util/encoding.go [hunk 23]
```diff
@@ -0,0 +1,239 @@
...
+	// The num should be [1,6765). Actually once the num is larger than or equal to 987,
+	// the efficiency of Fibonacci Encoding would be no better than 'WriteUint16'.
+	if num <= 0 || num >= fibLookup[fibLen-1]+fibLookup[fibLen-2] {
+		return fibEncodeNumOutOfRangeErr
+	}
+	// Binary Search to find the largest fibonacci number less than or equal to num.
+	lo, hi := 2, fibLen-1
+	for lo < hi {
+		mid := (lo + hi) / 2
+		if num >= fibonacci(mid) && num < fibonacci(mid+1) {
```
Comment (hhhjort) [inline util/encoding.go]: These calls to `fibonacci()` can also be replaced with array lookups to optimize thingsa little.

File: util/encoding.go (hunk 21 above)
Comment (hhhjort) [reply util/encoding.go:214]: The current name will be an issue when TCFv2 is brought into the project and the original Int range is implemented. The documentation refers to them as `Range (Fibonacci)` and `Range (Int)`, so the current name could confuse people looking for the `Range (Fibonacci)` implementation. This is why the corresponding read function is called `ReadFibonacciRange()`.

File: encode.go [hunk 24]
```diff
@@ -0,0 +1,63 @@
...
+	// Generate int range object.
+	intRange := new(util.IntRange)
+	var prevID constants.SectionID = -1
+	for _, sec := range sections {
+		id := sec.GetID()
+		if id == prevID {
+			return "", duplicatedSectionErr
+		}
+		if prevID+1 == id {
+			intRange.Range[len(intRange.Range)-1].EndID = uint16(id)
```
Comment (guscarreon) [inline encode.go:51]: If `id` equals zero in the first loop of this `for` statement, we'll be referencing `Range` at position `-1`. The following corner case indeed makes `Encode(sections []Section)` panic:
```
 16   type gppEncodeTestData struct {
 17       description string
 18       sections    []Section
 19       expected    string
 20   }
 21  
    + type zeroSection struct{}
    + 
    + func (s zeroSection) Encode(b bool) []byte {
    +     return nil
    + }
    + 
    + func (s zeroSection) GetID() constants.SectionID {
    +     return constants.SectionID(0)
    + }
    + 
    + func (s zeroSection) GetValue() string {
    +     return "string"
    + }
    + 
 22   var testData = []gppEncodeTestData{
    +     {
    +         description: "Corner case of a section whose GetID returns 0",
    +         sections:    []Section{zeroSection{}},
    +     },
 23       {
 24           description: "USPCA GPP string encoding",
 25           expected:    "DBABh4A~BlgWEYCY.QA~BSFgmiU",
 26           sections: []Section{
 27               uspca.USPCA{
 28                   CoreSegment: uspca.USPCACoreSegment{
 29                       Version:                     1,
 30                       SaleOptOutNotice:            2,
 31                       SharingOptOutNotice:         1,
 32                       SensitiveDataLimitUseNotice: 1,
 33 *--251 lines: SaleOptOut:                  2,-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
284       },
285   }
286  
287   func TestEncode(t *testing.T) {
288       for _, test := range testData {
289           result, err := Encode(test.sections)
encode_test.go
```
Can we add a check?

**Comment by bsardo** [comment]
Hi @duduchristian. It looks like we're close here. If you could address the remaining comment when you get a chance, that would be great. Thanks!

**Review by hhhjort — DISMISSED** [review]
LGTM

File: util/encoding.go [hunk 25]
```diff
@@ -0,0 +1,243 @@
...
+	"errors"
+)
+
+var (
+	fibEncodeNumOutOfRangeErr = errors.New("the number to be encoded is out of range")
+	fibEncodeInvalidRange     = errors.New("the range is invalid")
+)
+
+func getByteSlice() []byte {
+	// Most string to be encoded are less than 8 bytes.
```
Comment (guscarreon) [inline util/encoding.go]: Nitpick: "Most strings"

File: encode_test.go [hunk 26]
```diff
@@ -0,0 +1,378 @@
...
+					TargetedAdvertisingOptOut:       1,
+					SensitiveDataProcessing: []byte{
+						1, 2, 0, 0, 2, 1, 2, 2,
+					},
+					KnownChildSensitiveDataConsents: []byte{0},
+					MspaCoveredTransaction:          2,
+					MspaOptOutOptionMode:            1,
+					MspaServiceProviderMode:         1,
+				},
+				SectionID: 9,
```
Comment (guscarreon) [inline encode_test.go]: Nitpick: can we set to it's `constants.SectionUSPVA` value instead?

**Review by guscarreon — COMMENTED** [review]
Looks pretty good to me. I just left two very minor comments

**Review by guscarreon — APPROVED** [review]
LGTM

**Review by hhhjort — APPROVED** [review]
